File: runtime/Cpp/runtime/src/dfa/DFASerializer.cpp
/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

#include <sstream>
#include <vector>
#include <string>
#include <cstddef>
#include "dfa/DFA.h"
#include "Vocabulary.h"

#include "dfa/DFASerializer.h"

using namespace antlr4::dfa;

DFASerializer::DFASerializer(const DFA *dfa, const Vocabulary &vocabulary) : _dfa(dfa), _vocabulary(vocabulary) {
}

std::string DFASerializer::toString() const {
  if (_dfa->s0 == nullptr) {
    return "";
  }

  std::stringstream ss;
  std::vector<DFAState *> states = _dfa->getStates();
  for (auto *s : states) {
    for (size_t i = 0; i < s->edgeCount(); i++) {
      DFAState *t = s->getEdge(i);
      if (t != nullptr && t->stateNumber != INT32_MAX) {
        ss << getStateString(s);
        std::string label = getEdgeLabel(i);
        ss << "-" << label << "->" << getStateString(t) << "\n";
      }
    }
  }

  return ss.str();
}

std::string DFASerializer::getEdgeLabel(size_t i) const {
  // Edges are indexed by t + 1 (EOF in slot 0), so shift back to the token
  // type; i == 0 wraps to Token::EOF, which the vocabulary displays as "EOF".
  return _vocabulary.getDisplayName(i - 1);
}

std::string DFASerializer::getStateString(DFAState *s) const {
  size_t n = s->stateNumber;

  const std::string baseStateStr = std::string(s->isAcceptState ? ":" : "") + "s" + std::to_string(n) +
    (s->requiresFullContext ? "^" : "");

  if (s->isAcceptState) {
    if (!s->predicates.empty()) {
      std::string buf;
      for (size_t i = 0; i < s->predicates.size(); i++) {
        buf.append(s->predicates[i].toString());
      }
      return baseStateStr + "=>" + buf;
    } else {
      return baseStateStr + "=>" + std::to_string(s->prediction);
    }
  } else {
    return baseStateStr;
  }
}

File: runtime/Swift/Sources/Antlr4/dfa/DFASerializer.swift
/// 
/// Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
/// Use of this file is governed by the BSD 3-clause license that
/// can be found in the LICENSE.txt file in the project root.
/// 


/// 
/// A DFA walker that knows how to dump them to serialized strings.
/// 

public class DFASerializer: CustomStringConvertible {
    private let dfa: DFA
    private let vocabulary: Vocabulary

    public init(_ dfa: DFA, _ vocabulary: Vocabulary) {
        self.dfa = dfa
        self.vocabulary = vocabulary
    }

    public var description: String {
        if dfa.s0 == nil {
            return ""
        }
        var buf = ""
        let states = dfa.getStates()
        for s in states {
            guard let edges = s.edges else {
                continue
            }
            for (i, t) in edges.enumerated() {
                guard let t = t, t.stateNumber != Int.max else {
                    continue
                }
                let edgeLabel = getEdgeLabel(i)
                buf += getStateString(s)
                buf += "-\(edgeLabel)->"
                buf += getStateString(t)
                buf += "\n"
            }
        }

        return buf
    }

    internal func getEdgeLabel(_ i: Int) -> String {
        return vocabulary.getDisplayName(i - 1)
    }


    internal func getStateString(_ s: DFAState) -> String {
        let n = s.stateNumber

        let s1 = s.isAcceptState ? ":" : ""
        let s2 = s.requiresFullContext ? "^" : ""
        let baseStateStr = s1 + "s" + String(n) + s2
        if s.isAcceptState {
            if let predicates = s.predicates {
                return baseStateStr + "=>\(predicates)"
            } else {
                return baseStateStr + "=>\(s.prediction)"
            }
        } else {
            return baseStateStr
        }
    }
}
